In [1]:
!pip install git+https://github.com/jgy4/stable-baselines3

  Cloning https://github.com/jgy4/stable-baselines3 to /tmp/pip-req-build-0ozestei
  Running command git clone -q https://github.com/jgy4/stable-baselines3 /tmp/pip-req-build-0ozestei
  Created wheel for stable-baselines3: filename=stable_baselines3-1.4.1a0-py3-none-any.whl size=165193 sha256=f46a48d2c8c49bdbdfc1a979a3f83b9afa836183bc3d7584892d997ee3177871
  Stored in directory: /tmp/pip-ephem-wheel-cache-lgvi8y6y/wheels/34/ec/25/916523bacfe8c2bb53f600149a93109d04e1fcb9be04ba3089
Successfully built stable-baselines3


In [2]:
import gym
import numpy as np
import matplotlib.pyplot as plt
from stable_baselines3 import A2C
from stable_baselines3.common.evaluation import evaluate_policy

## N-Step Advantage Model

In [3]:
env = gym.make('CartPole-v1')
eval_env = gym.make('CartPole-v1')
model = A2C('MlpPolicy', env, verbose=1, n_step=True)
model.learn(total_timesteps=int(1e5))

Using cpu device
Wrapping the env with a `Monitor` wrapper
Wrapping the env in a DummyVecEnv.
------------------------------------
| rollout/              |          |
|    ep_len_mean        | 23.4     |
|    ep_rew_mean        | 23.4     |
| time/                 |          |
|    fps                | 535      |
|    iterations         | 100      |
|    time_elapsed       | 0        |
|    total_timesteps    | 500      |
| train/                |          |
|    entropy_loss       | -0.686   |
|    explained_variance | -0.074   |
|    learning_rate      | 0.0007   |
|    n_updates          | 99       |
|    policy_loss        | -1.11    |
|    value_loss         | 16.4     |
------------------------------------
------------------------------------
| rollout/              |          |
|    ep_len_mean        | 28.5     |
|    ep_rew_mean        | 28.5     |
| time/                 |          |
|    fps                | 617      |
|    iterations         | 200      |
|    time_elapsed 

In [4]:
mean_reward, std_reward = evaluate_policy(model, eval_env, n_eval_episodes=10, deterministic=True)

print(f"mean_reward={mean_reward:.2f} +/- {std_reward}")

/usr/local/lib/python3.7/dist-packages/stable_baselines3/common/evaluation.py:69: UserWarning: Evaluation environment is not wrapped with a ``Monitor`` wrapper. This may result in reporting modified episode lengths and rewards, if other wrappers happen to modify these. Consider wrapping environment first with ``Monitor`` wrapper.
  UserWarning,


mean_reward=500.00 +/- 0.0


## Vanilla Model

In [5]:
env = gym.make('CartPole-v1')
eval_env = gym.make('CartPole-v1')
model = A2C('MlpPolicy', env, verbose=1, n_step=False, gae_lambda=0)
model.learn(total_timesteps=int(1e5))

Using cpu device
Wrapping the env with a `Monitor` wrapper
Wrapping the env in a DummyVecEnv.
------------------------------------
| rollout/              |          |
|    ep_len_mean        | 9.6      |
|    ep_rew_mean        | 9.6      |
| time/                 |          |
|    fps                | 860      |
|    iterations         | 100      |
|    time_elapsed       | 0        |
|    total_timesteps    | 500      |
| train/                |          |
|    entropy_loss       | -0.0503  |
|    explained_variance | 0.44     |
|    learning_rate      | 0.0007   |
|    n_updates          | 99       |
|    policy_loss        | 0.0118   |
|    value_loss         | 3.08     |
------------------------------------
-------------------------------------
| rollout/              |           |
|    ep_len_mean        | 9.51      |
|    ep_rew_mean        | 9.51      |
| time/                 |           |
|    fps                | 878       |
|    iterations         | 200       |
|    time_e

In [6]:
mean_reward, std_reward = evaluate_policy(model, eval_env, n_eval_episodes=10, deterministic=True)

print(f"mean_reward={mean_reward:.2f} +/- {std_reward}")

/usr/local/lib/python3.7/dist-packages/stable_baselines3/common/evaluation.py:69: UserWarning: Evaluation environment is not wrapped with a ``Monitor`` wrapper. This may result in reporting modified episode lengths and rewards, if other wrappers happen to modify these. Consider wrapping environment first with ``Monitor`` wrapper.
  UserWarning,


mean_reward=70.00 +/- 21.066561181170506


## Monte Carlo

In [7]:
env = gym.make('CartPole-v1')
eval_env = gym.make('CartPole-v1')
model = A2C('MlpPolicy', env, verbose=1, n_step=False, gae_lambda=1)
model.learn(total_timesteps=int(1e5))

Using cpu device
Wrapping the env with a `Monitor` wrapper
Wrapping the env in a DummyVecEnv.
------------------------------------
| rollout/              |          |
|    ep_len_mean        | 22.2     |
|    ep_rew_mean        | 22.2     |
| time/                 |          |
|    fps                | 858      |
|    iterations         | 100      |
|    time_elapsed       | 0        |
|    total_timesteps    | 500      |
| train/                |          |
|    entropy_loss       | -0.69    |
|    explained_variance | 0.0308   |
|    learning_rate      | 0.0007   |
|    n_updates          | 99       |
|    policy_loss        | 1.76     |
|    value_loss         | 8.81     |
------------------------------------
------------------------------------
| rollout/              |          |
|    ep_len_mean        | 25.6     |
|    ep_rew_mean        | 25.6     |
| time/                 |          |
|    fps                | 870      |
|    iterations         | 200      |
|    time_elapsed 

In [8]:
mean_reward, std_reward = evaluate_policy(model, eval_env, n_eval_episodes=10, deterministic=True)

print(f"mean_reward={mean_reward:.2f} +/- {std_reward}")

/usr/local/lib/python3.7/dist-packages/stable_baselines3/common/evaluation.py:69: UserWarning: Evaluation environment is not wrapped with a ``Monitor`` wrapper. This may result in reporting modified episode lengths and rewards, if other wrappers happen to modify these. Consider wrapping environment first with ``Monitor`` wrapper.
  UserWarning,


mean_reward=263.50 +/- 8.958236433584458


### Default GAE Without N-Step

In [9]:
env = gym.make('CartPole-v1')
eval_env = gym.make('CartPole-v1')
model = A2C('MlpPolicy', env, verbose=1, n_step=False)
model.learn(total_timesteps=int(1e5))

Using cpu device
Wrapping the env with a `Monitor` wrapper
Wrapping the env in a DummyVecEnv.
------------------------------------
| rollout/              |          |
|    ep_len_mean        | 43.7     |
|    ep_rew_mean        | 43.7     |
| time/                 |          |
|    fps                | 879      |
|    iterations         | 100      |
|    time_elapsed       | 0        |
|    total_timesteps    | 500      |
| train/                |          |
|    entropy_loss       | -0.568   |
|    explained_variance | -0.042   |
|    learning_rate      | 0.0007   |
|    n_updates          | 99       |
|    policy_loss        | 0.619    |
|    value_loss         | 3.56     |
------------------------------------
------------------------------------
| rollout/              |          |
|    ep_len_mean        | 41.7     |
|    ep_rew_mean        | 41.7     |
| time/                 |          |
|    fps                | 876      |
|    iterations         | 200      |
|    time_elapsed 

In [10]:
mean_reward, std_reward = evaluate_policy(model, eval_env, n_eval_episodes=10, deterministic=True)

print(f"mean_reward={mean_reward:.2f} +/- {std_reward}")

/usr/local/lib/python3.7/dist-packages/stable_baselines3/common/evaluation.py:69: UserWarning: Evaluation environment is not wrapped with a ``Monitor`` wrapper. This may result in reporting modified episode lengths and rewards, if other wrappers happen to modify these. Consider wrapping environment first with ``Monitor`` wrapper.
  UserWarning,


mean_reward=500.00 +/- 0.0


In comparing the models we found the following results:

*   N-Step Mean Reward: 500
*   Vanilla Mean Reward: 70
*   Monte Carlo Mean Reward: 263.5
*   GAE Mean Reward: 500

The N-Step and GAE models performed the best with the highest mean rewards. The Monte Carlo model performed slightly worse with a mean reward of 263.5. The Vanilla model performed the worst with a mean reward of only 70.